### ISSUE TITLE
NumberFormatException: For input string: "17.8"

### ISSUE THREAD

**Issue by laeubi** [eclipse-jdt/eclipse.jdt#23]
I'm recently getting this when perform a java search after a workspace crash:
```
java.lang.NumberFormatException: For input string: "17.8"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at org.eclipse.jdt.internal.compiler.util.CtSym.getReleaseCode(CtSym.java:429)
	at org.eclipse.jdt.internal.core.builder.ClasspathJrtWithReleaseOption.initialize(ClasspathJrtWithReleaseOption.java:109)
	at org.eclipse.jdt.internal.core.builder.ClasspathJrtWithReleaseOption.<init>(ClasspathJrtWithReleaseOption.java:82)
	at org.eclipse.jdt.internal.core.builder.ClasspathLocation.forJrtSystem(ClasspathLocation.java:157)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.mapToClassPathLocation(JavaSearchNameEnvironment.java:267)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.computeClasspathLocations(JavaSearchNameEnvironment.java:200)
	at org.eclipse.jdt.internal.core.search.matching.JavaSearchNameEnvironment.<init>(JavaSearchNameEnvironment.java:94)
	at org.eclipse.jdt.internal.core.search.matching.IndexBasedJavaSearchEnvironment.create(IndexBasedJavaSearchEnvironment.java:27)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.initialize(MatchLocator.java:1239)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.locateMatches(MatchLocator.java:1277)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.locateMatches(MatchLocator.java:1390)
	at org.eclipse.jdt.internal.core.search.matching.MatchLocator.locateMatches(MatchLocator.java:1509)
	at org.eclipse.jdt.internal.core.search.JavaSearchParticipant.locateMatches(JavaSearchParticipant.java:135)
	at org.eclipse.jdt.internal.core.search.BasicSearchEngine.findMatches(BasicSearchEngine.java:251)
	at org.eclipse.jdt.internal.core.search.BasicSearchEngine.search(BasicSearchEngine.java:602)
	at org.eclipse.jdt.core.search.SearchEngine.search(SearchEngine.java:670)
	at org.eclipse.jdt.internal.ui.search.JavaSearchQuery.run(JavaSearchQuery.java:178)
	at org.eclipse.search2.internal.ui.InternalSearchUI$InternalSearchJob.run(InternalSearchUI.java:94)
```
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

**Comment by laeubi** [comment]
Any chance to get a feedback here? Can I delete a cached file or something? This nearly makes everything unusable, reference search, finding implementations and even save actions complain here no with this error... :-\

**Comment by laeubi** [comment]
Also save actions fail:
```
Java Model Exception: java.lang.NumberFormatException: For input string: "17.7"
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:326)
	at org.eclipse.jdt.internal.core.CompilationUnit.buildStructure(CompilationUnit.java:193)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:266)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:597)
	at org.eclipse.jdt.internal.core.CompilationUnit.makeConsistent(CompilationUnit.java:1142)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:173)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:806)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1319)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:132)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:94)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:91)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:158)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:76)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:207)
Caused by: java.lang.NumberFormatException: For input string: "17.7"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at org.eclipse.jdt.internal.compiler.impl.CompilerOptions.releaseToJDKLevel(CompilerOptions.java:854)
	at org.eclipse.jdt.internal.core.JarPackageFragmentRoot.computeChildren(JarPackageFragmentRoot.java:135)
	at org.eclipse.jdt.internal.core.JavaProjectElementInfo.initializePackageNames(JavaProjectElementInfo.java:276)
	at org.eclipse.jdt.internal.core.JavaProjectElementInfo.getProjectCache(JavaProjectElementInfo.java:244)
	at org.eclipse.jdt.internal.core.JavaProject.getProjectCache(JavaProject.java:2478)
	at org.eclipse.jdt.internal.core.JavaProjectElementInfo.newNameLookup(JavaProjectElementInfo.java:333)
	at org.eclipse.jdt.internal.core.JavaProject.newNameLookup(JavaProject.java:2897)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.<init>(SearchableEnvironment.java:104)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.<init>(SearchableEnvironment.java:134)
	at org.eclipse.jdt.internal.core.CancelableNameEnvironment.<init>(CancelableNameEnvironment.java:32)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:255)
	... 18 more
Caused by: java.lang.NumberFormatException: For input string: "17.7"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at org.eclipse.jdt.internal.compiler.impl.CompilerOptions.releaseToJDKLevel(CompilerOptions.java:854)
	at org.eclipse.jdt.internal.core.JarPackageFragmentRoot.computeChildren(JarPackageFragmentRoot.java:135)
	at org.eclipse.jdt.internal.core.JavaProjectElementInfo.initializePackageNames(JavaProjectElementInfo.java:276)
	at org.eclipse.jdt.internal.core.JavaProjectElementInfo.getProjectCache(JavaProjectElementInfo.java:244)
	at org.eclipse.jdt.internal.core.JavaProject.getProjectCache(JavaProject.java:2478)
	at org.eclipse.jdt.internal.core.JavaProjectElementInfo.newNameLookup(JavaProjectElementInfo.java:333)
	at org.eclipse.jdt.internal.core.JavaProject.newNameLookup(JavaProject.java:2897)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.<init>(SearchableEnvironment.java:104)
	at org.eclipse.jdt.internal.core.SearchableEnvironment.<init>(SearchableEnvironment.java:134)
	at org.eclipse.jdt.internal.core.CancelableNameEnvironment.<init>(CancelableNameEnvironment.java:32)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:255)
	at org.eclipse.jdt.internal.core.CompilationUnit.buildStructure(CompilationUnit.java:193)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:266)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:597)
	at org.eclipse.jdt.internal.core.CompilationUnit.makeConsistent(CompilationUnit.java:1142)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:173)
	at 
org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:806)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1319)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:132)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:94)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:91)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:158)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:76)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:207)
```

**Comment by laeubi** [comment]
What helps for a while is deleting the index and restarting eclipse.

**Comment by iloveeclipse** [comment]
> What helps for a while is deleting the index and restarting eclipse.

So it is reproducible and you see it again and again?
Which JVM is used, which Eclipse version ? Anything else is "special" in your case (since we don't see similar reports)?

**Comment by laeubi** [comment]
> So it is reproducible and you see it again and again?

Not really reproducible in a way that I can give any steps, but I see this in two installations the past days, it *seem* it might be triggered by a Eclipse-crash. I'm not sure what these code actually do, but maybe if it is connected to any cached data, maybe it would be good to trow away the cache in cases there is any error? I just saw

> org.eclipse.jdt.internal.core.JavaProject.getProjectCache(JavaProject.java:2478)

in the stack and thought it could be related to a broken cache...

> Eclipse IDE for RCP and RAP Developers (includes Incubating components)
> 
> Version: 2022-06 (4.24.0)
> Build id: 20220609-1112

Used VM: jdk-11.0.8

**Comment by iloveeclipse** [comment]
So in both cases your Eclipse crashed before, right?

**Comment by laeubi** [comment]
> So in both cases your Eclipse crashed before, right?

I think the first time it occurred was where i need to kill the process because of https://github.com/eclipse-platform/eclipse.platform.debug/issues/26.

Then yesterday a Tycho workspace crashed while debugging and then I see the error there as well, but its just a guess that it might be related.

**Comment by iloveeclipse** [comment]
I wonder if you can grep in all your `.settings/org.eclipse.jdt.core.prefs` files in all your projects in the workspace if you can find this value set: `org.eclipse.jdt.core.compiler.compliance=17.8` ?

If there is no one, check `.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.jdt.core.prefs` in the corresponding workspace, but I guess it must be broken project settings file.

I assume you've managed to merge broken preference value somehow.

BTW, your first stack is NOT from 4.24 final, but from some earlier build, it doesn't match 4.24 master.

**Comment by laeubi** [comment]
> BTW, your first stack is NOT from 4.24 final, but from some earlier build, it doesn't match 4.24 master.

Yep I tried to upgrade and hoped it would be gone then... Should I do the search now or wait until the problem occurs again? As mentioned one can clear the cache and then it seem to work again.

**Comment by iloveeclipse** [comment]
No, this broken value is **saved** in some of your setting files, JDT reads it every time again, it must be manually fixed.

**Comment by laeubi** [comment]
I found some files containing:

```
eclipse.preferences.version=1
org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.8
org.eclipse.jdt.core.compiler.compliance=17.8
org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
org.eclipse.jdt.core.compiler.problem.reportPreviewFeatures=ignore
org.eclipse.jdt.core.compiler.release=disabled
org.eclipse.jdt.core.compiler.source=1.8
```

**Comment by iloveeclipse** [comment]
> I found some files containing

The most interesting question: **how** your files got to this state? Manual/automatic merge?

**Comment by laeubi** [comment]
> .metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.jdt.core.prefs

Looks like this:
```
eclipse.preferences.version=1
org.eclipse.jdt.core.classpathVariable.ECLIPSE_HOME=<path to workspace>/.metadata/.plugins/org.eclipse.pde.core/.bundle_pool
org.eclipse.jdt.core.classpathVariable.JRE_LIB=<path to java>/zulu11.37.19-ca-fx-jdk11.0.6-linux_x64/lib/jrt-fs.jar
org.eclipse.jdt.core.classpathVariable.JRE_SRC=<path to java>/zulu11.37.19-ca-fx-jdk11.0.6-linux_x64/lib/src.zip
org.eclipse.jdt.core.classpathVariable.JRE_SRCROOT=
org.eclipse.jdt.core.classpathVariable.JUNIT_HOME=<userhome>/.p2/pool/plugins/org.junit_4.13.2.v20211018-1956.jar
org.eclipse.jdt.core.classpathVariable.M2_REPO=<userhome>/.m2/repository
org.eclipse.jdt.core.classpathVariable.TARGET_PLATFORM_BUNDLE_POOL=<path to workspace>/.metadata/.plugins/org.eclipse.pde.core/.bundle_pool
org.eclipse.jdt.core.codeComplete.visibilityCheck=enabled
org.eclipse.jdt.core.compiler.codegen.inlineJsrBytecode=enabled
org.eclipse.jdt.core.compiler.codegen.targetPlatform=11
org.eclipse.jdt.core.compiler.compliance=11
org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
org.eclipse.jdt.core.compiler.problem.invalidJavadoc=warning
org.eclipse.jdt.core.compiler.problem.invalidJavadocTags=enabled
org.eclipse.jdt.core.compiler.problem.invalidJavadocTagsDeprecatedRef=enabled
org.eclipse.jdt.core.compiler.problem.invalidJavadocTagsNotVisibleRef=enabled
org.eclipse.jdt.core.compiler.problem.missingJavadocComments=warning
org.eclipse.jdt.core.compiler.problem.missingJavadocTagDescription=all_standard_tags
org.eclipse.jdt.core.compiler.problem.missingJavadocTags=warning
org.eclipse.jdt.core.compiler.release=enabled
org.eclipse.jdt.core.compiler.source=11
```

**Comment by laeubi** [comment]
> Manual/automatic merge?

The seem not managed by the CVS, but are inside a maven project so maybe maven derived wrong values here?

**Comment by laeubi** [comment]
If I choose Compare > Local history eclipse says there is no history for this files, so it seems I have not modified them manually.

**Comment by iloveeclipse** [comment]
Whoever generated "org.eclipse.jdt.core.compiler.compliance=17.8" did surely something wrong.

**Comment by laeubi** [comment]
@mickaelistria do you know if m2e is setting this property?

Is there any chance for JDT to transform this in an error marker instead of failing completely with a strange error dialog?

**Comment by laeubi** [comment]
The only relation i could find is that the appendix seems to be related by the other settings e.g. I found this file:

```
org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.5
org.eclipse.jdt.core.compiler.compliance=17.5
org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
org.eclipse.jdt.core.compiler.problem.reportPreviewFeatures=ignore
org.eclipse.jdt.core.compiler.release=disabled
org.eclipse.jdt.core.compiler.source=1.5
```

So `targetPlatform=1.5` / `compiler.source=1.5` --> `compiler.compliance=17.5`

other file:
```
eclipse.preferences.version=1
org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.8
org.eclipse.jdt.core.compiler.compliance=17.8
org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
org.eclipse.jdt.core.compiler.problem.reportPreviewFeatures=ignore
org.eclipse.jdt.core.compiler.release=disabled
org.eclipse.jdt.core.compiler.source=1.8
```

`codegen.targetPlatform=1.8` / `compiler.source=1.8` --> `compiler.compliance=17.8`

**Comment by laeubi** [comment]
Same here:
```
eclipse.preferences.version=1
org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.7
org.eclipse.jdt.core.compiler.compliance=17.7
org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
org.eclipse.jdt.core.compiler.problem.reportPreviewFeatures=ignore
org.eclipse.jdt.core.compiler.release=disabled
org.eclipse.jdt.core.compiler.source=1.7
```

**Comment by iloveeclipse** [comment]
> Is there any chance for JDT to transform this in an error marker instead of failing completely with a strange error dialog?

Marker probably not, but proper error at the first time reading this probably possible. The problem is, in the place where options are read, there are no semantic checks for the possible content, and values are used in various places where the source of the problem isn't clear anymore (is this project or workspace option broken?)

**Comment by laeubi** [comment]
Error log would be fine I think at least I would have never thought to search the preference files :-)

So something like invalid "org.eclipse.jdt.core.compiler.compliance" with value "whatever"  instead of the current "NumberFormatException: For input string: "17.8"" would probably help already :-)

Beside that, tha bad value seems to be generated by m2e but the code is a bit hard to follow why... I think the problem is that there are cases where a dot version is allowed (e.g. 1.8 / 1.7 / ...) and others where an integer is expsected ( 8 / 7 / ...) while maven always seem to return integers so there is some code to transform the value...

**Comment by laeubi** [comment]
> we should just take care of it before calling releaseToJDKLevel(). Notice how versionToJdkLevel() is prepared for this format.

So you mean the value is not wrong but probably the code is not aware of it?

**Comment by jarthana** [comment]
In this case (for Java 17), the value is wrong. But in the past, we did have compliance versions that could be of this format, like 1.5 etc.
